### PR TITLE
HBASE-23107 Avoid temp byte array creation when doing cacheDataOnWrite

### PR DESCRIPTION
**Pull request by chenxu14**

### CODE REVIEW

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   1m 51s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| :blue_heart: |  mvndep  |   0m 41s |  Maven dependency ordering for branch  |
| :green_heart: |  mvninstall  |   6m 40s |  master passed  |
| :green_heart: |  compile  |   1m 23s |  master passed  |
| :green_heart: |  checkstyle  |   1m 59s |  master passed  |
| :green_heart: |  shadedjars  |   5m 27s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   0m 56s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 50s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   5m 44s |  master passed  |
||| _ Patch Compile Tests _ |
| :blue_heart: |  mvndep  |   0m 16s |  Maven dependency ordering for patch  |
| :green_heart: |  mvninstall  |   7m 12s |  the patch passed  |
| :green_heart: |  compile  |   1m 33s |  the patch passed  |
| :green_heart: |  javac  |   1m 33s |  the patch passed  |
| :green_heart: |  checkstyle  |   0m 29s |  The patch passed checkstyle in hbase-common  |
| :green_heart: |  checkstyle  |   1m 44s |  hbase-server: The patch generated 0 new + 51 unchanged - 1 fixed = 51 total (was 52)  |
| :broken_heart: |  whitespace  |   0m  0s |  The patch has 1 line(s) that end in whitespace. Use git apply --whitespace=fix <<patch_file>>. Refer https://git-scm.com/docs/git-apply  |
| :green_heart: |  shadedjars  |   5m 40s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  17m 20s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 57s |  the patch passed  |
| :green_heart: |  findbugs  |   5m 27s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  |   2m 56s |  hbase-common in the patch passed.  |
| :broken_heart: |  unit  | 295m 18s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   0m 56s |  The patch does not generate ASF License warnings.  |
|  |   | 371m 48s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.master.balancer.TestStochasticLoadBalancerRegionReplicaSameHosts |
|   | hadoop.hbase.master.TestMasterShutdown |
|   | hadoop.hbase.client.TestMobSnapshotCloneIndependence |
|   | hadoop.hbase.master.balancer.TestStochasticLoadBalancerRegionReplicaHighReplication |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/678 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux cea581fd47ab 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-678/out/precommit/personality/provided.sh |
| git revision | master / 2ebdcbc3b0 |
| Default Java | 1.8.0_181 |
| whitespace | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/1/artifact/out/whitespace-eol.txt |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/1/testReport/ |
| Max. process+thread count | 5299 (vs. ulimit of 10000) |
| modules | C: hbase-common hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/1/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 33s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| :blue_heart: |  mvndep  |   0m 38s |  Maven dependency ordering for branch  |
| :green_heart: |  mvninstall  |   5m 15s |  master passed  |
| :green_heart: |  compile  |   1m 21s |  master passed  |
| :green_heart: |  checkstyle  |   1m 46s |  master passed  |
| :green_heart: |  shadedjars  |   4m 39s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   1m  1s |  master passed  |
| :blue_heart: |  spotbugs  |   4m  5s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 51s |  master passed  |
||| _ Patch Compile Tests _ |
| :blue_heart: |  mvndep  |   0m 15s |  Maven dependency ordering for patch  |
| :green_heart: |  mvninstall  |   4m 54s |  the patch passed  |
| :green_heart: |  compile  |   1m 18s |  the patch passed  |
| :green_heart: |  javac  |   1m 18s |  the patch passed  |
| :green_heart: |  checkstyle  |   0m 27s |  The patch passed checkstyle in hbase-common  |
| :green_heart: |  checkstyle  |   1m 17s |  hbase-server: The patch generated 0 new + 51 unchanged - 1 fixed = 51 total (was 52)  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 38s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  15m 35s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   0m 58s |  the patch passed  |
| :green_heart: |  findbugs  |   5m  1s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  |   3m  6s |  hbase-common in the patch passed.  |
| :broken_heart: |  unit  | 163m 19s |  hbase-server in the patch failed.  |
| :green_heart: |  asflicense  |   1m 11s |  The patch does not generate ASF License warnings.  |
|  |   | 228m 44s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.master.assignment.TestOpenRegionProcedureHang |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/678 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux e81ffae7ac33 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-678/out/precommit/personality/provided.sh |
| git revision | master / 7ee6d59ef8 |
| Default Java | 1.8.0_181 |
| unit | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/2/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/2/testReport/ |
| Max. process+thread count | 4839 (vs. ulimit of 10000) |
| modules | C: hbase-common hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/2/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by virajjasani** [comment]
Since impl of cloneUncompressedBufferWithHeader() is to be changed, do you think it is better to keep the method signature same?

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 34s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| :blue_heart: |  mvndep  |   0m 44s |  Maven dependency ordering for branch  |
| :green_heart: |  mvninstall  |   5m 34s |  master passed  |
| :green_heart: |  compile  |   1m 19s |  master passed  |
| :green_heart: |  checkstyle  |   1m 46s |  master passed  |
| :green_heart: |  shadedjars  |   4m 39s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   1m  4s |  master passed  |
| :blue_heart: |  spotbugs  |   4m  5s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   4m 50s |  master passed  |
||| _ Patch Compile Tests _ |
| :blue_heart: |  mvndep  |   0m 15s |  Maven dependency ordering for patch  |
| :green_heart: |  mvninstall  |   4m 58s |  the patch passed  |
| :green_heart: |  compile  |   1m 19s |  the patch passed  |
| :green_heart: |  javac  |   1m 19s |  the patch passed  |
| :green_heart: |  checkstyle  |   0m 26s |  The patch passed checkstyle in hbase-common  |
| :green_heart: |  checkstyle  |   1m 17s |  hbase-server: The patch generated 0 new + 51 unchanged - 1 fixed = 51 total (was 52)  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   4m 37s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  15m 48s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   1m  1s |  the patch passed  |
| :green_heart: |  findbugs  |   4m 58s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  |   3m  7s |  hbase-common in the patch passed.  |
| :green_heart: |  unit  | 158m 33s |  hbase-server in the patch passed.  |
| :green_heart: |  asflicense  |   1m  8s |  The patch does not generate ASF License warnings.  |
|  |   | 224m 27s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.2 Server=19.03.2 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/678 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux d246644b10d1 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-678/out/precommit/personality/provided.sh |
| git revision | master / 10180e232e |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/3/testReport/ |
| Max. process+thread count | 4647 (vs. ulimit of 10000) |
| modules | C: hbase-common hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/3/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by chenxu14** [comment]
Hi @virajjasani ,do you mean we'd better not change cloneUncompressedBufferWithHeader to cloneUncompressedBuffWithHeader? Already done so. Thanks for the review.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileWriterImpl.java [hunk 1]
```diff
@@ -548,6 +547,8 @@ private void doCacheOnWrite(long offset) {
       HFileBlock cacheFormatBlock = blockWriter.getBlockForCaching(cacheConf);
       cache.cacheBlock(new BlockCacheKey(name, offset, true, cacheFormatBlock.getBlockType()),
           cacheFormatBlock);
+      // refCnt will auto increase when block add to Cache, see RAMCache#putIfAbsent
+      cacheFormatBlock.release();
```
Comment (openinx) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileWriterImpl.java]: Better to release the cacheFormatBlock in the finally block ?  Say even if we encounter some exception when cache#cacheBlock, the byte buffer allocated from ByteBuffAllocator can still have the chance to release ( return back to the pool ) ..

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java [hunk 2]
```diff
@@ -857,11 +859,17 @@ EncodingState getEncodingState() {
      * @param dataBlockEncoder data block encoding algorithm to use
      */
     public Writer(HFileDataBlockEncoder dataBlockEncoder, HFileContext fileContext) {
+      this(dataBlockEncoder, fileContext, ByteBuffAllocator.HEAP);
```
Comment (openinx) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java]: Seems this constructor is only used for UT now ?  If so please mark it as VisiableForTesting and say only for testing purpose.. Or Just move it out of the Writer class (put it in UT class )?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java [hunk 3]
```diff
@@ -1170,19 +1190,19 @@ int blockSizeWritten() {
...
+      ByteBuff bytebuff = allocator.allocate(baosInMemory.size());
+      baosInMemory.toByteBuff(bytebuff);
       int numBytes = (int) ChecksumUtil.numBytes(
           onDiskBlockBytesWithHeader.size(),
           fileContext.getBytesPerChecksum());
-      putHeader(uncompressedBlockBytesWithHeader, 0,
-        onDiskBlockBytesWithHeader.size() + numBytes,
-        baosInMemory.size(), onDiskBlockBytesWithHeader.size());
-      return ByteBuffer.wrap(uncompressedBlockBytesWithHeader);
+      putHeader(bytebuff, onDiskBlockBytesWithHeader.size() + numBytes,
```
Comment (openinx) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java]: Just curious:   we put the bytes from baosInMemory  into the newly allocated byteBuff ( start from 0 pos ),  then overwrite it by putting the header (still start from 0 pos..), any problem ?

Comment (chenxu14) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java]: The bytes in baosInMemory has no header info, just reserved the header space, so we should overwrite it in putHeader. @openinx

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| :blue_heart: |  reexec  |   0m 51s |  Docker mode activated.  |
||| _ Prechecks _ |
| :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ master Compile Tests _ |
| :blue_heart: |  mvndep  |   0m 42s |  Maven dependency ordering for branch  |
| :green_heart: |  mvninstall  |   7m 21s |  master passed  |
| :green_heart: |  compile  |   1m 36s |  master passed  |
| :green_heart: |  checkstyle  |   2m  9s |  master passed  |
| :green_heart: |  shadedjars  |   5m 58s |  branch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  javadoc  |   1m  8s |  master passed  |
| :blue_heart: |  spotbugs  |   4m 56s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| :green_heart: |  findbugs  |   5m 53s |  master passed  |
||| _ Patch Compile Tests _ |
| :blue_heart: |  mvndep  |   0m 16s |  Maven dependency ordering for patch  |
| :green_heart: |  mvninstall  |   6m 49s |  the patch passed  |
| :green_heart: |  compile  |   1m 38s |  the patch passed  |
| :green_heart: |  javac  |   1m 38s |  the patch passed  |
| :green_heart: |  checkstyle  |   0m 30s |  The patch passed checkstyle in hbase-common  |
| :green_heart: |  checkstyle  |   1m 37s |  hbase-server: The patch generated 0 new + 51 unchanged - 1 fixed = 51 total (was 52)  |
| :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| :green_heart: |  shadedjars  |   5m 51s |  patch has no errors when building our shaded downstream artifacts.  |
| :green_heart: |  hadoopcheck  |  19m 57s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2 or 3.1.2.  |
| :green_heart: |  javadoc  |   1m  7s |  the patch passed  |
| :green_heart: |  findbugs  |   5m 38s |  the patch passed  |
||| _ Other Tests _ |
| :green_heart: |  unit  |   3m  7s |  hbase-common in the patch passed.  |
| :green_heart: |  unit  | 170m 14s |  hbase-server in the patch passed.  |
| :green_heart: |  asflicense  |   0m 58s |  The patch does not generate ASF License warnings.  |
|  |   | 251m 43s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.3 Server=19.03.3 base: https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/678 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 673179e62674 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-slave/workspace/HBase-PreCommit-GitHub-PR_PR-678/out/precommit/personality/provided.sh |
| git revision | master / 1aee5f0552 |
| Default Java | 1.8.0_181 |
|  Test Results | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/4/testReport/ |
| Max. process+thread count | 5063 (vs. ulimit of 10000) |
| modules | C: hbase-common hbase-server U: . |
| Console output | https://builds.apache.org/job/HBase-PreCommit-GitHub-PR/job/PR-678/4/console |
| versions | git=2.11.0 maven=2018-06-17T18:33:14Z) findbugs=3.1.11 |
| Powered by | Apache Yetus 0.11.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java [hunk 4]
```diff
@@ -1012,6 +1021,18 @@ private void putHeader(byte[] dest, int offset, int onDiskSize,
       Bytes.putInt(dest, offset, onDiskDataSize);
     }
 
+    private void putHeader(ByteBuff buff, int onDiskSize,
```
Comment (openinx) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java]: Seems we can abstract the two putHeader into a single one ?   the byte[] dest can also be regarded as a ByteBuff ?

Comment (chenxu14) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java]: you mean wrap the dest like this?
    **ByteBuff.wrap(ByteBuffer.wrap(dest))**
Will create extra ByteBuffer and ByteBuff instance, not friendly to YGC?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileWriterImpl.java [hunk 5]
```diff
@@ -546,8 +545,13 @@ private void writeInlineBlocks(boolean closing) throws IOException {
...
     cacheConf.getBlockCache().ifPresent(cache -> {
       HFileBlock cacheFormatBlock = blockWriter.getBlockForCaching(cacheConf);
-      cache.cacheBlock(new BlockCacheKey(name, offset, true, cacheFormatBlock.getBlockType()),
-          cacheFormatBlock);
+      try {
+        cache.cacheBlock(new BlockCacheKey(name, offset, true, cacheFormatBlock.getBlockType()),
+            cacheFormatBlock);
+      } finally {
+        // refCnt will auto increase when block add to Cache, see RAMCache#putIfAbsent
+        cacheFormatBlock.release();
```
Comment (ramkrish86) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileWriterImpl.java]: Pardon me here. So where does the initial increment happen? RamCache#putIfAbsent will do a retain anyway.

Comment (chenxu14) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileWriterImpl.java]: When ByteBuffAllocator#allocate called, the returned ByteBuff will have an RefCnt, it’s initial value is 1 @ramkrish86

Comment (ramkrish86) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileWriterImpl.java]: Ya got it. Thanks.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java (hunk 4 above)
Comment (openinx) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/HFileBlock.java]: That's OK.

**Pull request merged**